### PR TITLE
SAK-45616 add Proctorio capabilities to Samigo

### PR DESCRIPTION
**Pull request by ottenhoff**

### CODE REVIEW

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    try {
+      HibernateCallback hcb = (Session s) -> {
+        Query q = s.getNamedQuery(QUERY_GET_URLS_FOR_ASSESSMENT);
+        q.setParameter("publishedId", assessmentId);
+        return q.list();
+      };
+    } catch (DataAccessException e) {
+      log.error("Failed to retrieve URLs from SecureDelivery table for assessmentId={}", assessmentId);
+    }
+    return null;
```
Comment (bbailla2) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java:47]: This method queries then does nothing with the result

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      HibernateCallback hcb = (Session s) -> {
+        Query q = s.getNamedQuery(QUERY_GET_URLS_FOR_ASSESSMENT_AND_USER);
+        q.setParameter("publishedId", "assessmentId");
+        q.setParameter("agentId", agentId);
+        return q.list();
+      };
+    } catch (DataAccessException e) {
+      log.error("Failed to retrieve URLs from SecureDelivery table for assessmentId={}", assessmentId);
+    }
+    return null;
```
Comment (bbailla2) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java:61]: Same as above

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/SecureDeliveryProctorio.java [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+			"calculatorbasic", "calculatorsci", "whiteboard"
+	};
+
+	private static String proctorioKey;
+	private static String proctorioSecret;
+	private static String proctorioUrl;
+	private static String proctorioEnabled;
+	private Cache<String, String> urlCache;
+
+	private UserDirectoryService userDirectoryService = ComponentManager.get(UserDirectoryService.class);
```
Comment (ern) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/SecureDeliveryProctorio.java:77]: new code should look at using SpringBeanAutowiringSupport

Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/SecureDeliveryProctorio.java:77]: i definitely tried using the AutoWired annotations

File: samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/grading/SecureDeliveryData.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.sakaiproject.tool.assessment.data.dao.grading;
+
+import lombok.*;
```
Comment (adrianfish) [inline samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/grading/SecureDeliveryData.java:20]: I'd expand this. There are only three imports and I think we frown upon wildcards, unless we're bringing in loads.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java (hunk 1 above)
Comment (adrianfish) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java:47]: It doesn't query, just creates the callback ready for use in a hibernatetemplate call. It just needs the hibernate template piece after.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/PersistenceService.java [hunk 5]
```diff
@@ -247,8 +246,13 @@ public void setExtendedTimeFacade(ExtendedTimeFacade extendedTimeFacade) {
 	public ExtendedTimeFacade getExtendedTimeFacade() {
 		return extendedTimeFacade;
 	}
-}
-
 
+	public void setSecureDeliveryFacadeQueries(SecureDeliveryFacadeQueriesAPI secureDeliveryFacadeQueries) {
+		this.secureDeliveryFacadeQueries = secureDeliveryFacadeQueries;
```
Comment (adrianfish) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/PersistenceService.java:251]: indentation

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java [hunk 6]
```diff
@@ -105,10 +109,29 @@ public void init() {
 	 * @returns true if at least one secure delivery module implementation is available.
 	 */
 	public boolean isSecureDeliveryAvaliable() {
-		
-		return secureDeliveryModules.size() > 0;
+		for ( Map.Entry<String, SecureDeliveryModuleIfc> entry : secureDeliveryModules.entrySet() ) {
+			if (entry.getValue().isEnabled()) {
+				return true;
+			}
+		}
```
Comment (adrianfish) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:116]: indentation whacked out

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java [hunk 7]
```diff
@@ -186,7 +209,7 @@ public PhaseStatus validatePhase(String moduleId, Phase phase, PublishedAssessme
 		
 		SecureDeliveryModuleIfc module = secureDeliveryModules.get( moduleId );
 		
-		if ( moduleId == null || NONE_ID.equals( moduleId ) || module == null || !module.isEnabled() )
+		if ( moduleId == null || NONE_ID.equals( moduleId ) || module == null || !module.isEnabled(assessment.getPublishedAssessmentId()) )
```
Comment (adrianfish) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:212]: Maybe use stringutils.isblank on moduleId?

**Review by adrianfish — COMMENTED** [review]
Just pedantic stuff around spacing, indentation, the usual Samigo stuff. You're always going to get null back from those hibernate callback calls, too, I think. You need a template call after creating the callback.

File: samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/SecureDeliveryModuleIfc.java [hunk 8]
```diff
@@ -110,4 +110,38 @@
...
+	/**
+	 * Returns an absolute URL to an alternative location to take the assessment.
+	 * For example, a cloud proctoring company could use this URL to embed the Samigo assessment in an iframe.
+	 * Using a default empty return to avoid breaking legacy code.
+	 * @param assessmentId
+	 * @param uid
+	 * @return
+	 */
+	default String getAlternativeDeliveryUrl (Long assessmentId, String uid) {
+		return "";
```
[the remark this answers is not in the document]
Comment (ottenhoff) [reply samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/SecureDeliveryModuleIfc.java:123]: it's a proper tab mate!

File: samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/shared/api/assessment/SecureDeliveryServiceAPI.java [hunk 9]
```diff
@@ -148,4 +149,11 @@
 	 * @return the reference. null if the module is not avaliable or if the module rejected the context 
 	 */
 	public SecureDeliveryModuleIfc getModuleReference( String moduleId, Object context );
+
+	public String getAlternativeDeliveryUrl( String moduleId, Long assessmentId, String uid );
+
+	public String getInstructorReviewUrl( String moduleId, Long assessmentId, String studentId );
+
+	public boolean isSecureDeliveryAvaliable(Long publishedAssessmentId);
```
[the remark this answers is not in the document]
Comment (ottenhoff) [reply samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/shared/api/assessment/SecureDeliveryServiceAPI.java:157]: done

File: samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/grading/SecureDeliveryData.java (hunk 4 above)
Comment (ottenhoff) [reply samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/grading/SecureDeliveryData.java:20]: done agree

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/PersistenceService.java (hunk 5 above)
Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/PersistenceService.java:251]: another proper tab

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java (hunk 6 above)
Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:116]: you seem to hate tabs. they're fine

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java (hunk 7 above)
Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:212]: really don't want to refactor samigo code if not needed

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java [hunk 10]
```diff
@@ -348,6 +403,18 @@ private void handlePlugin( String secureDeliveryPlugin ) {
 	
 		try
 		{
+			// This is a built-in integration with no additional JAR file
+			if (secureDeliveryPlugin.equalsIgnoreCase("proctorio")) {
+				ApplicationContext appCtx = new AnnotationConfigApplicationContext(SecureDeliveryProctorio.class);
```
[the remark this answers is not in the document]
Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:408]: I burned a dozen hours trying to use existing context

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java (hunk 2 above)
Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java]: thanks

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java (hunk 1 above)
Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/SecureDeliveryFacadeQueries.java]: cheers, fixed

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/TotalScoreListener.java [hunk 11]
```diff
@@ -573,6 +575,14 @@ public void prepareAgentResult(PublishedAssessmentData p, Iterator iter, List ag
 	
 	TotalScoresBean bean = (TotalScoresBean) ContextUtil.lookupBean("totalScores");
 	Map agentResultsByAssessmentGradingIdMap = new HashMap();
+
+	SecureDeliveryServiceAPI secureDelivery = SamigoApiFactory.getInstance().getSecureDeliveryServiceAPI();
+	String moduleId = null;
+    if ( secureDelivery.isSecureDeliveryAvaliable() ) {
```
[the remark this answers is not in the document]
Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/TotalScoreListener.java:581]: replaces these tabs with spaces

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/TotalScoreListener.java [hunk 12]
```diff
@@ -596,6 +606,12 @@ public void prepareAgentResult(PublishedAssessmentData p, Iterator iter, List ag
     	  if (gdata.getForGrade()) {
     		  results.setTotalAutoScore(gdata.getTotalAutoScore().toString());
     		  results.setForGrade(Boolean.TRUE);
+
+    		  // If the student took the assessment, then provide a view of their proctoring (if available)
+    		  if (StringUtils.isNotBlank(moduleId)) {
+    			  results.setAlternativeInstructorReviewUrl( secureDelivery.getInstructorReviewUrl(moduleId, new Long(p.getPublishedAssessmentId()), gdata.getAgentId()) );
```
[the remark this answers is not in the document]
Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/TotalScoreListener.java:611]: matches spacing scheme from above

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/select/SelectActionListener.java [hunk 13]
```diff
@@ -139,10 +138,10 @@ public void processAction(ActionEvent ae) throws
     }
     
     // filter out the one that the given user do not have right to access
-    List takeableList = getTakeableList(publishedAssessmentList,  h, updatedAssessmentNeedResubmitList, updatedAssessmentList);
+    List<PublishedAssessmentFacade> takeableList = getTakeableList(publishedAssessmentList,  h, updatedAssessmentNeedResubmitList, updatedAssessmentList);
```
[the remark this answers is not in the document]
Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/select/SelectActionListener.java:141]: done

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/select/SelectActionListener.java [hunk 14]
```diff
@@ -427,18 +426,27 @@ else if ("2".equals(select.getDisplayAllAssessments())) {
...
-    	
-    	HttpServletRequest request = (HttpServletRequest) FacesContext.getCurrentInstance().getExternalContext().getRequest();
-    	select.setSecureDeliveryHTMLFragments( secureDelivery.getInitialHTMLFragments(request, new ResourceLoader().getLocale() ) );
+        HttpServletRequest request = (HttpServletRequest) FacesContext.getCurrentInstance().getExternalContext().getRequest();
+        select.setSecureDeliveryHTMLFragments( secureDelivery.getInitialHTMLFragments(request, new ResourceLoader().getLocale() ) );
+
+        for (DeliveryBeanie db : takeablePublishedList) {
+            // We have to refetch the published assessment because the hash above doesn't have the metadata
+            PublishedAssessmentFacade paf = publishedAssessmentService.getPublishedAssessment(db.getAssessmentId());
+            final String moduleId = paf.getAssessmentMetaDataByLabel( SecureDeliveryServiceAPI.MODULE_KEY );
```
[the remark this answers is not in the document]
Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/select/SelectActionListener.java:440]: done

**Pull request closed without merging**